// lik/dist/lik.limitedarray.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class LimitedArray {
    constructor(limitArg) {
        this.array = [];
        this.arrayLimit = limitArg;
    }
    addOne(objectArg) {
        this.array.unshift(objectArg);
        if (this.array.length > this.arrayLimit) {
            this.array.length = this.arrayLimit;
        }
    }
    addMany(objectArrayArg) {
        for (let objectArg of objectArrayArg) {
            this.addOne(objectArg);
        }
    }
    setLimit(limitArg) {
        this.arrayLimit = limitArg;
        if (this.array.length > this.arrayLimit) {
            this.array.length = this.arrayLimit;
        }
    }
    getAverage() {
        if (typeof this.array[0] === 'number') {
            let sum = 0;
            for (let localNumber of this.array) {
                let localNumberAny = localNumber;
                sum = sum + localNumberAny;
            }
            return sum / this.array.length;
        }
        else {
            return null;
        }
    }
}
exports.LimitedArray = LimitedArray;
