// smartstring/dist/smartstring.git.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/* ---------------------------------------------- *
 * ------------------ classes ------------------- *
 * ---------------------------------------------- */
class GitRepo {
    constructor(stringArg, tokenArg) {
        let regexMatches = gitRegex(stringArg);
        this.host = regexMatches[1];
        this.user = regexMatches[2];
        this.repo = regexMatches[3];
        this.accessToken = tokenArg;
        this.sshUrl = gitLink(this.host, this.user, this.repo, this.accessToken, 'ssh');
        this.httpsUrl = gitLink(this.host, this.user, this.repo, this.accessToken, 'https');
    }
}
exports.GitRepo = GitRepo;
/* ---------------------------------------------- *
 * ------------------ helpers ------------------- *
 * ---------------------------------------------- */
const gitRegex = function (stringArg) {
    const regexString = /([a-zA-Z0-9\-\.]*)(?:\/|\:)([a-zA-Z0-9\-\.]*)(?:\/)([a-zA-Z0-9\-\.]*)(?:\.git)/;
    let regexMatches = regexString.exec(stringArg);
    return regexMatches;
};
const gitLink = function (hostArg, userArg, repoArg, tokenArg = '', linkTypeArg) {
    let returnString;
    if (tokenArg !== '') {
        tokenArg = tokenArg + '@';
    }
    switch (linkTypeArg) {
        case 'https':
            returnString = 'https://' +
                tokenArg + hostArg + '/' + userArg + '/' + repoArg + '.git';
            break;
        case 'ssh':
            returnString = 'git@' +
                hostArg + ':' + userArg + '/' + repoArg + '.git';
            break;
        default:
            console.error('Link Type ' + linkTypeArg + ' not known');
            break;
    }
    return returnString;
};
