// smartenv/dist/smartenv.envhelpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEnvString = function () {
    if (typeof window !== 'undefined') {
        return 'browser';
    }
    else if (typeof process !== 'undefined') {
        return 'node';
    }
};
exports.getUserAgentString = function () {
    if (exports.isBrowser()) {
        return navigator.userAgent;
    }
    else {
        return undefined;
    }
};
exports.isNode = function () {
    return exports.getEnvString() === 'node';
};
exports.getNodeVersion = function () {
    return process.version;
};
exports.isBrowser = function () {
    return !exports.isNode();
};
exports.isCI = function () {
    if (process.env.CI) {
        return true;
    }
    else {
        return false;
    }
    ;
};
exports.isC9 = function () {
    if (process.env.C9_HOSTNAME) {
        return true;
    }
    else {
        return false;
    }
};
exports.isTravis = function () {
    if (process.env.TRAVIS) {
        return true;
    }
    else {
        return false;
    }
    ;
};
